### PR TITLE
Feature: "accessible handout mode"

### PR DESCRIPTION
**Pull request by salbeira**
Adds a feature to the "a11y" plugin:

The user can click a button in the navigation menu to view the presentation as a single scrollable page instead of a slide deck. This is done so assistive technology can parse and navigate through the presentation as a single document.

In addition this allows us to sunset the handout feature of decker as all of our reveal.js plugins are not activated inside handouts, causing them to lack a lot of features and content conversions.

Also I heared some people like to scroll through presentations on their smartphone rather than swipe around.

The feature is implemented by "kidnapping" the slides into a fake reveal element. This is done so our css rules responsible for layouting are kept the same. The "kidnapping" is necessary because if the slides are kept inside the reveal container, reveal itself randomly re-establishes its layouting and resets all attribute changes that are necessary to enable the scrolling view.

Intersection Observers are used to load and unload iframes and videos that scroll in and out of view.

The feedback plugin is modified so it does not reload the "currentSlide" questions if the handout mode is enabled and instead takes the data from the currently "in view" slide while the handout mode is active.

### CODE REVIEW

File: bin/build.ps1 [hunk 1]
```diff
@@ -34,17 +34,12 @@ if (-Not $skiptemplates) {
 <# Cleanup of old files #> 
 Write-Host "Cleaning before new build" -ForegroundColor Green
 & stack clean
-if (test-path "$deckerdir\resource\decker\support\vendor") {
-    Remove-Item "$deckerdir\resource\decker\support\vendor" -Recurse -Force -ErrorAction Continue
```
Comment (mbotsch) [inline bin/build.ps1:38]: What does this have to do with "accesible handout mode"?

File: bin/build.ps1 [hunk 2]
```diff
@@ -34,17 +34,12 @@ if (-Not $skiptemplates) {
...
-}
 if (test-path "$deckerdir\public"){
     Remove-Item "$deckerdir\public" -Recurse -Force -ErrorAction Ignore 
 }
 
 
 Write-Host "Starting build of standalone binary" -ForegroundColor Green
-& git submodule update --init
-& .\third-party\vendor.ps1
```
Comment (mbotsch) [inline bin/build.ps1:47]: same here: shouldn't this branch just add the handout mode?

File: decker.yaml [hunk 3]
```diff
@@ -25,6 +25,8 @@ static-resources:
 # resource-pack: exe:mario
 # resource-pack: exe:mono
 
+resource-pack: exe:tudo
```
Comment (mbotsch) [inline decker.yaml:28]: do not enforce tudo style for all decks in this project (even though it is the nicest style)

File: resource/decker/support/plugins/a11y/a11y.js [hunk 4]
```diff
@@ -65,9 +70,33 @@ const Plugin = {
   id: "a11y",
   init: (reveal) => {
     Reveal = reveal;
-    fixTabsByInert();
+    // This may no longer be necessary if we clearly recommend the handout mode to people using assistive technology
+    // fixTabsByInert();
     addFlyingFocusCallbacks();
     addCustomSpacebarHandler();
```
Comment (mbotsch) [inline resource/decker/support/plugins/a11y/a11y.js:76]: if this mode does not automatically trigger the handout mode, then it should also disable all autoplay videos (and add controls)

File: resource/decker/support/plugins/decker/decker.js [hunk 5]
```diff
@@ -15,6 +15,10 @@ function onStart(deck) {
 
   deck.addEventListener("ready", () => {
     if (!printMode) {
+      totalSlides = deck.getTotalSlides();
+      console.log(
+        "[decker.js plugin] The deck has " + totalSlides + " total slides."
```
Comment (mbotsch) [inline resource/decker/support/plugins/decker/decker.js:20]: remove this debug output in order to not fill up the console

File: resource/decker/support/plugins/handout/handout.js [hunk 6]
```diff
@@ -0,0 +1,370 @@
...
+    slide.inert = false;
+    slide.hidden = false;
+    slide.removeAttribute("aria-hidden");
+  }
+}
+
+function findAllRealSlides(slidesElement, slideList) {
+  for (const child of slidesElement.childNodes) {
+    if (child.tagName !== "SECTION") continue;
+    if (child.classList.contains("stack")) {
```
Comment (mbotsch) [inline resource/decker/support/plugins/handout/handout.js:22]: slidesElement.querySelectorAll("section.stack")

File: resource/decker/support/plugins/handout/handout.js [hunk 7]
```diff
@@ -0,0 +1,370 @@
...
+      for (const subchild of child.childNodes) {
+        if (subchild.tagName !== "SECTION") continue;
```
Comment (mbotsch) [inline resource/decker/support/plugins/handout/handout.js:24]: getElementsByTagName("section")

File: resource/decker/support/plugins/handout/handout.js [hunk 8]
```diff
@@ -0,0 +1,370 @@
...
+        subchild.previousParent = child;
+        const indices = Reveal.getIndices(subchild);
+        if (indices && indices.h) {
+          child.dataset.hIndex = indices.h;
```
Comment (mbotsch) [inline resource/decker/support/plugins/handout/handout.js:28]: why does child get the index of subchild?

File: resource/decker/support/plugins/live-captioning/live-captioning.js [hunk 9]
```diff
@@ -65,9 +65,8 @@ function getLanguageCode(string) {
  * Plugin class to be registered so the activation-(decker-)button is at the right place.
  */
 class LiveCaptioning {
-  constructor() {
-    this.reveal = undefined;
-    this.id = "live-captioning";
+  constructor(reveal) {
```
Comment (mbotsch) [inline resource/decker/support/plugins/live-captioning/live-captioning.js:68]: why is this plugin involved in this branch?

File: resource/decker/support/plugins/menu/menu.js [hunk 10]
```diff
@@ -516,6 +508,7 @@ class SlideMenu {
     this.menu.close_button = container.querySelector(
       "#decker-menu-close-button"
     );
+    this.menu.a11y_button = container.querySelector("#decker-menu-a11y-button");
```
Comment (mbotsch) [inline resource/decker/support/plugins/menu/menu.js:511]: why does this button show up explicitly, when it is dynamically added in the handout plugin?

File: bin/build.ps1 (hunk 1 above)
Comment (salbeira) [reply bin/build.ps1:38]: Nothing, I just happened to mention this while working inside the branch and accidentally added it here - we should cherry pick that out into develop and master because it deletes the vendor directory that has been statically added to the support directory.

File: decker.yaml (hunk 3 above)
Comment (salbeira) [reply decker.yaml]: Commented it out again

File: resource/decker/support/plugins/decker/decker.js (hunk 5 above)
Comment (salbeira) [reply resource/decker/support/plugins/decker/decker.js]: Done.

File: resource/decker/support/plugins/handout/handout.js (hunk 7 above)
Comment (salbeira) [reply resource/decker/support/plugins/handout/handout.js]: The whole codeflow of this part of the plugin has changed and should be using getElementsByTagName or proper querySelectors now.

File: resource/decker/support/plugins/handout/handout.js (hunk 8 above)
Comment (salbeira) [reply resource/decker/support/plugins/handout/handout.js]: That was a typo. Fixed.

File: resource/decker/support/plugins/live-captioning/live-captioning.js (hunk 9 above)
Comment (salbeira) [reply resource/decker/support/plugins/live-captioning/live-captioning.js]: It seems I forgot to add these files to the speaker notes fix branch and it got added in an add -u later in this branch. Sorry for that, we can cherry pick this out into develop.

File: resource/decker/support/plugins/menu/menu.js (hunk 10 above)
Comment (salbeira) [reply resource/decker/support/plugins/menu/menu.js]: That was a leftover from my first trys to implement the mode. Removed it.

File: resource/decker/support/plugins/menu/menu.js [hunk 11]
```diff
@@ -554,6 +547,22 @@ class SlideMenu {
...
+    button.classList.add("fa-button", "fas", icon);
+    button.title = title;
+    button.setAttribute("aria-label", title);
+    button.addEventListener("click", callback);
+    const menuHeaderButtons = document.getElementsByClassName(
+      "menu-header-buttons"
+    );
+    if (menuHeaderButtons.length > 0) {
+      const container = menuHeaderButtons[0];
+      container.appendChild(button);
```
[the remark this answers is not in the document]
Comment (salbeira) [reply resource/decker/support/plugins/menu/menu.js]: Should be implemented as described.

**Comment by mbotsch** [comment]
Look's good, thanks! I noticed two more things that do not work:
- Keyboard navigation by cursor up/down, page up/down, home/end does not work in handout mode.
- On tablets and phones one cannot scroll up and down by touch input (hence one cannot scroll at all).

**Comment by mbotsch** [comment]
I fixed the background images and the touch scrolling issue.
The keyboard navigation is still open.

**Comment by monofon** [comment]
Very nice! 

So far,  I found one thing that does not seem to work: background interactive iframes.

```markdown
# ![](/apps/image-zoom.html){.frame .interactive}
```

Produces an empty slide. The iFrame is nowhere to be seen. The iframe app works as expected if not used on the slide background.

**Comment by salbeira** [comment]
The keyboard one is a bit of a doozy: The default keyboard behaviour can be restored by disabling reveal's keyboard configuration. None of our custom keyboard commands are available then though.

I am adding a custom keyboard event handler that scrolls the viewport if we are in handout mode.

@monofon Can you list what media elements are available as background elements? (Before you wrote this I thought images and videos are the height of technological advancements in background technology)

**Comment by monofon** [comment]
Standard Reveal feature: https://revealjs.com/backgrounds/

**Comment by mbotsch** [comment]
I implemented background videos for handout mode. I guess that background iframes should work in a similar fashion.

@monofon are you using background iframes for your slides? I didn't even know that those existed...

**Comment by monofon** [comment]
Yes, i do.

**Comment by mbotsch** [comment]
I just noticed that background videos and background iframes should use `data-src` instead of `src` and the intersection observer should then load the content when the slide becomes visible.

File: resource/decker/support/plugins/handout/handout.js [hunk 12]
```diff
@@ -0,0 +1,571 @@
...
+    .forEach((section) => {
+      const div = document.createElement("div");
+      div.classList.add("background");
+      const url = section.getAttribute("data-background-image");
+      const size = section.getAttribute("data-background-size");
+      const position = section.getAttribute("data-background-position");
+      const repeat = section.getAttribute("data-background-repeat");
+      div.style.backgroundImage = `url("${url}")`;
+      div.style.backgroundSize = size;
+      div.style.backgroundPosition = position;
```
Comment (mbotsch) [inline resource/decker/support/plugins/handout/handout.js:63]: check whether attributes have been defined by user before setting them

File: resource/decker/support/plugins/handout/handout.js [hunk 13]
```diff
@@ -0,0 +1,571 @@
...
+    });
+
+  // setup background videos
+  slidesElement
+    .querySelectorAll("section[data-background-video]")
+    .forEach((section) => {
+      const video = document.createElement("video");
+      const url = section.getAttribute("data-background-video");
+      const loop = section.getAttribute("data-background-video-loop");
+      video.dataset.src = url;
```
Comment (mbotsch) [inline resource/decker/support/plugins/handout/handout.js:75]: add muted attribute

File: resource/decker/support/plugins/handout/handout.js [hunk 14]
```diff
@@ -0,0 +1,571 @@
...
+  // remove background videos
+  fakeRevealContainer.querySelectorAll("video.background").forEach((video) => {
+    video.remove();
+  });
+  // setup background iframes
+  fakeRevealContainer
+    .querySelectorAll("iframe.background")
+    .forEach((iframe) => {
+      iframe.remove();
+    });
```
Comment (mbotsch) [inline resource/decker/support/plugins/handout/handout.js:147]: can be done in one loop with adjusted selector

File: resource/decker/support/plugins/handout/handout.js [hunk 15]
```diff
@@ -0,0 +1,571 @@
...
+  for (const slide of iterate) {
+    revealSlidesElement.appendChild(slide);
+  }
+  fakeRevealContainer.parentElement.insertBefore(
+    revealContainer,
+    fakeRevealContainer.nextSibling
+  );
+  detachWindowEventListeners();
+  /* Remove the fake container from the DOM */
+  fakeRevealContainer.parentElement.removeChild(fakeRevealContainer);
```
Comment (mbotsch) [inline resource/decker/support/plugins/handout/handout.js:159]: why not fakeRevealContainer.remove()?

File: resource/decker/support/plugins/handout/handout.js [hunk 16]
```diff
@@ -0,0 +1,571 @@
...
+  fakeRevealContainer.addEventListener("scroll", determineMostVisibleSlide);
+}
+
+/**
+ * Create an intersection observer that observes all video, audio and iframe
+ * elements to only load their src if they are close to being visible.
+ * @returns Nothing
+ */
+function createSRCIntersectionObserver() {
+  if (srcIntersectionObserver) {
```
Comment (mbotsch) [inline resource/decker/support/plugins/handout/handout.js:312]: bug: background videos/iframes are added dynamically each time handout mode is activated. if oberservers are not re-created these new elements are not observed (hence not loaded).

File: resource/decker/support/plugins/handout/handout.js [hunk 17]
```diff
@@ -0,0 +1,571 @@
...
+      return;
+    }
+    entries.forEach((entry) => {
+      if (entry.isIntersecting) {
+        if (!entry.target.src && entry.target.dataset.src) {
+          entry.target.src = entry.target.dataset.src;
+        }
+      } else {
+        if (entry.target.src) {
+          entry.target.dataset.src = entry.target.src;
```
Comment (mbotsch) [inline resource/decker/support/plugins/handout/handout.js:337]: should not be necessary

File: resource/decker/support/plugins/handout/handout.js [hunk 18]
```diff
@@ -0,0 +1,571 @@
...
+  }
+  const audios = fakeRevealContainer.getElementsByTagName("AUDIO");
+  for (const audio of audios) {
+    srcIntersectionObserver.observe(audio);
+  }
+  const videos = fakeRevealContainer.getElementsByTagName("VIDEO");
+  for (const video of videos) {
+    srcIntersectionObserver.observe(video);
+  }
+}
```
Comment (mbotsch) [inline resource/decker/support/plugins/handout/handout.js:360]: again, adjust selector to do this in one loop

File: resource/decker/support/plugins/handout/handout.js [hunk 19]
```diff
@@ -0,0 +1,571 @@
...
+  fakeSlideContainer.style.transform = "scale(" + scale + ")";
+}
+
+let lockScrolling;
+
+/**
+ * Keyup listener to handle scrolling in handout mode
+ */
+function onWindowKeydown(event) {
+  if (lockScrolling) return;
```
Comment (mbotsch) [inline resource/decker/support/plugins/handout/handout.js:380]: scrolling shows random behavior on my example slides :-(

**Comment by mbotsch** [comment]
The ping-pong is getting inefficient. Let me give it a try...

**Comment by mbotsch** [comment]
I found some more quirks. Fixing them required that our UI anchors had to be move from the Reveal element to its parent (the reveal-viewport element). While this seem to be the more natural choice anyway, I had to make several CSS adjustments (presenter mode, (my) poll mode, whiteboard buttons).

Please test once more...
